### PR TITLE
Fixing Offline DQM Bug 

### PR DESCRIPTION
**Pull request by vkhristenko**
Fixing the bug reported here
https://hypernews.cern.ch/HyperNews/CMS/get/recoDevelopment/1396/1/3/1/2/1/1.html

This is only for Offline DQM!

VK

### CODE REVIEW

File: DQM/HcalMonitorTasks/src/HcalDigiMonitor.cc [hunk 1]
```diff
@@ -570,6 +570,8 @@ void HcalDigiMonitor::analyze(edm::Event const&e, edm::EventSetup const&s)
 		  i<=FEDNumbering::MAXHCALuTCAFEDID; i++) {
 	  if (i>FEDNumbering::MAXHCALFEDID && i<FEDNumbering::MINHCALuTCAFEDID)
 		  continue;
+	  if (i<1118)
```
Comment (slava77) [inline DQM/HcalMonitorTasks/src/HcalDigiMonitor.cc]: a comment here may be helpful (e.g. ``` skip uTCA non-HF FEDs``` ).

this hardcoded choice is not good for maintenance: at some point the HBHE uTCA will need to be monitored as well.

**Comment by slava77** [comment]
@vkhristenko 
What is the meaning of ``` This is only for Offline DQM! ``` ?
@danduggan @deguio are you in control of what goes in online DQM in a separate release builds?

**Comment by slava77** [comment]
@cmsbuild please test

File: DQM/HcalMonitorTasks/src/HcalDigiMonitor.cc (hunk 1 above)
Comment (slava77) [inline DQM/HcalMonitorTasks/src/HcalDigiMonitor.cc]: also, shouldn't this be ``` if (i > MAXHCALFEDID && i < 1118 ) ```
if I understand correctly, now only HF uTCA FEDs will pass.

**Comment by vkhristenko** [comment]
Hi Slava, so the situation is the following:
I'm currently rewritting HCAL DQM (Online + Offline) from scratch. As far as I've checked, and I might be mistaken, but I believe/hope I'm not, that is all I have to worry about (in terms of packages), Validation seems to use their own modules - therefore only Offline + Online.
Old modules will be completely depreciated starting December 14th(beginning of the Long Technical Stop). Physics Datataking stops at that point and I believe that will be the most optimal time to introduce the updates
if I commit and push (deploy) everything now I have to provide instructions + it will be completely confusing for the shifters/etc... and difficult for me. I do plan start committing newest development quite soon - ~middle of November. 
-> Why only for Offline - this question has been extensively discussed with Central DQM - we use something on top that has been running for several month now... both Dan and Federico are aware of this fact. 

-> Regarding why no comments - module is to be replaced as mentioned above. 

-> Regarding including my newly introduced if statement into the one just above - would of course be better - yes, the point is to skip uTCA/AMC13-based FEDs - no explanation. I just did it like that...
This issue first showed up in Online DQM and was fixed earlier today the same way and runs successfully now...

I hope that answers all the questions
VK

**Comment by danduggan** [comment]
@vkhristenko, I think @slava77's comments should be easy to add, could you please do so, then we can test + approve.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
with the previous commit I observe quite a few changes in the digi monitor.

E.g. in run1 data monitoring:
wf45 3_depth1_good_digis

the right-hand-side is "after the change in 2390933" it actually looks more correct.
(the "baseline" looks like the HF inputs are ignored as bad)
Does this mean that the HcalDigiMonitor is not changing the e-map depending on GT? (basically, hardcoded to what the latest detector readout is).

**Comment by vkhristenko** [comment]
So, first of @danduggan, you were totally right - I'm just too quick and it will make a difference - updated all 4 PRs

Regarding GT/emap - no, cause we directly look at Digi Collection, therefore we deal with whatever is unpacked only. And what gets unpacked sits in the config(I mean the settings for the Emap). Is the left plot from run1??? Cause in Run1 there was no uTCA readout, and if you apply the current mapping + unpacker - you will get no digis for HF - as there are no VME detIds in the emap.

1 thing I don't know exactly: I don't know which configuration file for dqm application you are running.
That might've been quite stupid, but I don't know...

VK

**Comment by slava77** [comment]
@vkhristenko you can reproduce the configuration by running
``` runTheMatrix.py -l 4.53 -j 0``` 
(remove -j0 to actually run; this will just create a configuration).
The left and right plots are for run1, which can be guessed from the run number.

**Comment by slava77** [comment]
@abdoulline

**Comment by vkhristenko** [comment]
I don't know why this change causes this kind of behavior, in particular for this plot.
The part that is changed has to do with extracting number of time samples from raw - that is how it was done for vme. And nothing with occupancies....

The changes that Slava asked were actually quite neccessary.
I'm clueless at this point

VK

**Comment by slava77** [comment]
@vkhristenko just in case: the plots I posted were with the first version of this PR (only changes in https://github.com/vkhristenko/cmssw/commit/2390933efd50525b7fcbb6a8a74f1e8a9133b8fc)
the tests with the latest version are not done yet.

**Comment by vkhristenko** [comment]
@slava77,  Yeah yeah, I understood the timeline.

**Comment by abdoulline** [comment]
Well, I'm guseesing what all this might mean...

(1)
In the Unpacker itself Ted Laird has added (half a year ago) HF uTCA FEDSs explicitly

http://cmslxr.fnal.gov/source/EventFilter/HcalRawToDigi/plugins/HcalRawToDigi.cc#0033
*because* interation range doesn't include uTCA HF FEDs.

(2)  here, in the initial 2390933 DQM change
https://github.com/cms-sw/cmssw/commit/2390933efd50525b7fcbb6a8a74f1e8a9133b8fc

                                 >731                                                           <1100   - mens non-HCAL
if (i>FEDNumbering::MAXHCALFEDID && i<FEDNumbering::MINHCALuTCAFEDID)
		  continue;
	  if (i<1118)   //   means drop all non-HF uTCA FEDs, what about uTCA HBHE?  
		  continue;

(3) emap ( HcalElectronicsMap_v7.05_hlt /  HcalElectronicsMap_v7.05_offline or earlier versions)  
should all be Run1 -compatible...

**Comment by vkhristenko** [comment]
@abdoulline, 
Regarding uTCA HBHE, they are 1100 up 1116. 
HF uTCA are 1118, 1120, 1122
Initial change isn't perfect - yes, skips all the non-HF-uTCA FEDs - fixed now.
However, that piece only affects getting number of time samples from raw. it doesn't look at anything else... it only works for vme....

Regarding the unpacker - I don't know....

VK

**Comment by abdoulline** [comment]
OK, thank you for providing HBHE uTCA FEDs (HF onse 1118/1120/1122 is 
excatly what's hardcoded in the "patch" of Ted Laird in the unpacker).

If you say
"However, that piece only affects getting number of time samples from raw. 
it doesn't look at anything else... it only works for vme...."

I'm confused by how it can affect Digi occupancy as reported by Slava...

S.

**Comment by abdoulline** [comment]
I've dumped/expanded config step3_RAW2DIGI_L1Reco_RECO_EI_PAT_ALCA_DQM.py
from
runTheMatrix.py -l 4.53 -j 0 

and can see that GT 'auto:run1_data_Fake' stands for 74X_dataRun1_v3, which in turn includes  HcalElectronicsMap_v7.05_offline with a (first) Run 1 IOV available...
https://cms-conddb.cern.ch/browser/#list/Prod/tags/HcalElectronicsMap_v7.05_offline
And I've dumped its payload and can see it does correspond to Run 1 emap (all-VME HF).
So emap itself should be OK for Run 1...

**Comment by danduggan** [comment]
@abdoulline and @vkhristenko, we just discussed this in the ORP, and if you're happy with the current implementation, then I'll sign off.

**Comment by vkhristenko** [comment]
I would like to wait for comparison of the tests, which is running now...

When it's done, Dan, unless Salavat has objections, I think that we can go ahead.

VK

**Comment by danduggan** [comment]
@vkhristenko perfect.

**Comment by abdoulline** [comment]
Looking at the (two) lines of code update, I have no objections.
Just cannot explain Slava's plots...

**Comment by vkhristenko** [comment]
I gotta say that after looking at the comparison (in particular since the newest comparison gives really good match for what didn't match after the initial commit), I gave it a closer look.
There are 2 more variables that do get affected in that part and get used later in the code - Therefore, my apology for that - I was too quick.

However, now it seems to be fine. I think we are ready (from Hcal side)

VK

**Comment by danduggan** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**